docs/current/guides/757394-use-service-containers.md

```diff
@@ -54,167 +54,290 @@ Service containers come with the following built-in features:
+:::info
+This example uses Redis's `SAVE` command to ensure data is synced. By default, Redis flushes data to disk periodically.
 :::
 
-When a service is bound to a container, it also conveys to any outputs of that container, such as files or directories. The service will be started whenever the output is used, so you can also do things like this:
+## Start and stop services
+
+Starting with Dagger v0.8.8, you can explicitly start and stop services in your pipelines.
 
-<Tabs groupId="language" className="embeds">
+<Tabs groupId="language">
 <TabItem value="Go">
 
-<Embed id="Qoljk5SEPuu" />
+TODO
 
 </TabItem>
 <TabItem value="Node.js">
 
-<Embed id="RhA4m6ji1js" />
+TODO
 
 </TabItem>
 <TabItem value="Python">
 
-<Embed id="WG0Pqr49pKK" />
+TODO
 
 </TabItem>
 </Tabs>
 
-## Understand the service lifecycle
+## Example: MariaDB database service for application tests
+
+The following example demonstrates service containers in action, by creating a MariaDB database service container for use in application unit/integration testing.
+
+The application used in this example is [Drupal](https://www.drupal.org/), a popular open-source PHP CMS. Drupal includes a large number of unit tests, including tests which require an active database connection. All Drupal 10.x tests are written and executed using the [PHPUnit](https://phpunit.de/) testing framework. Read more about [running PHPUnit tests in Drupal](https://www.drupal.org/docs/automated-testing/phpunit-in-drupal/running-phpunit-tests).
+
+<Tabs groupId="language">
+<TabItem value="Go">
+
+```go file=./snippets/use-services/use-db-service/main.go
+```
+
+</TabItem>
+<TabItem value="Node.js">
+
+```typescript file=./snippets/use-services/use-db-service/index.ts
+```
+
+</TabItem>
+<TabItem value="Python">
+
+```python file=./snippets/use-services/use-db-service/main.py
+```
+
+</TabItem>
+</Tabs>
+
+This example begins by creating a MariaDB service container and initializing a new MariaDB database. It then creates a Drupal container and installs required dependencies into it. Next, it adds a binding for the MariaDB service (`db`) in the Drupal container and sets a container environment variable (`SIMPLETEST_DB`) with the database DSN. Finally, it runs Drupal's kernel tests (which [require a database connection](https://www.drupal.org/docs/automated-testing/phpunit-in-drupal/running-phpunit-tests#non-unit-tests)) using PHPUnit and prints the test summary to the console.
+
+:::tip
+Explicitly specifying the service container port with `WithExposedPort()` (Go), `withExposedPort()` (Node.js) or `with_exposed_port()` (Python) is particularly important here. Without it, Dagger will start the service container and immediately allow access to service clients. With it, Dagger will wait for the service to be listening first.
+:::
+
+## Reference: How service binding works for container services
```

Comment:
@vito can you confirm no further changes are needed in this section?

Reply:
I think we need to change some of it. This part in particular:

> It's worth noting that services are just containers, and all containers in Dagger have run-exactly-once semantics. Concurrent runs of the same container synchronize and attach to the same run with progress/logs multiplexed to each caller. Canceling a run only interrupts the container process when all runs are canceled, so new clients can come and go throughout a service run.

Services are still "just containers" but now they run differently from regular containers. Instead of deduping them _globally_ across the Dagger engine like regular containers, they are only deduped _locally_ to each client. That means if you run the exact same service across separate clients/sessions, the service will run exactly once per client, whereas before it would run exactly once, period, and the clients would all be talking to the same service.

Suggested replacement:

> Services are based on containers, but they run a little differently. Whereas regular Containers in Dagger are de-duped across the entire engine, Service containers are only deduped within a Dagger client session. This means that if you run separate Dagger sessions that use the exact same services, they each will get their own "instance" of the service. This process is carefully tuned to preserve caching at each client call-site, while prohibiting "cross-talk" from one Dagger session's client to another Dagger session's service.

Looking ahead to the next paragraph:

> Run-exactly-once semantics are very convenient. You don't have to come up with names and maintain instances of services; they're content-addressed, so you just use them by value. You also don't have to manage the state of the service; you can just trust that it will be running when needed and stopped when not.

The latter part is still true, it's just not run-exactly-once anymore.

Suggested replacement:

> Content-addressed services are very convenient. You don't have to come up with names and maintain instances of services; just use them by value. You also don't have to manage the state of the service; you can just trust that it will be running when needed and stopped when not.

Reply:
Done